### PR TITLE
GH Actions CI

### PR DESCRIPTION
**Pull request by bradshjg**
Supercedes https://github.com/vcr/vcr/pull/881
Related to https://github.com/vcr/vcr/discussions/851

This represent a fairly drastic change in how tests are run, previously a single script `ci.sh` was responsible for running all the steps and appropriately timing and folding the relevant sections. This PR splits the CI steps into separate Actions steps, ~~but doesn't preserve at least one (probably important!) behavior: testing for warnings.~~ Edit: this now does check for warnings but the semantics might differ slightly from `ci.sh`.

~~It's not clear to me whether we want to dump all warnings in various steps to a file or create a similar abstraction as `ci.sh` had where warning checks were sorta built-in to the various steps. I found the indirection in `ci.sh` painful to follow and personally would probably prefer more explicit (if duplicated) `2> {some_warnings_file}` and then a step to crawl through those at the end, but I'm totally open to suggestions.~~ Edit: each test step dumps stderr to a file and then that file is checked for warnings in a later step.

Other potentially controversial things:

* `em-http-request` adds headers by default, and in the test suite there's very reasonable header tests that get broken by this behavior. I opted to update the request options in a way that omits those headers so the tests pass as written, but it's kinda weird and fragile but hey they started it! 😄 
*  I skipped a flaky multi-threaded test. See https://github.com/vcr/vcr/pull/814 for a little more context.

### CODE REVIEW

**Comment by bradshjg** [comment]
Sorry to open a new PR about this, but the other one was long enough ago that I completely re-imagined how I wanted to do this. I hope this makes sense to y'all too.

Things we talked about that I want to highlight:

> "let us experiment until we get to 🟢 green builds in this PR".

That's the spirit I went with, which means stuff like

> Can we add in "3.0" as well here?

I omitted because it is going to break and that probably deserves its own PR.

**Comment by krainboltgreene** [comment]
Okay, so what do we need to get this merged? If we have to strip stuff out that's fine, we can always work on readding those things, but I would love a list of things we're leaving behind.

**Comment by bradshjg** [comment]
> Okay, so what do we need to get this merged? If we have to strip stuff out that's fine, we can always work on readding those things, but I would love a list of things we're leaving behind.

The only thing I know that we're dropping with respect to testing is a test for the vcr library emitting warnings (with careful attention to exiting cleanly if only dependencies emit warnings), but I'd love a second set of 👀 on that.

I also would like to drop the entire `script/` folder before merge, but that's because I like my PRs 🟥 🟥  😄

**Comment by bradshjg** [comment]
@olleolleolle @krainboltgreene gentle bump to see if there's anything I can help to move this forward (and apologies for the direct mentions if they prove less than helpful) 😄

File: .github/workflows/test.yml
Comment (olleolleolle) [inline .github/workflows/test.yml:6]: I'm curious if this alias for default branch does the right thing. It's be cool.

**Review by olleolleolle — APPROVED** [review]
In order to keep the momentum, @bradshjg, let's merge this, which added a single defined skip, which can be investigated. Warnings are collected and analyzed. 

Em-http-related testing fixes for reproducibility could potentially be a separate thing, but I'm OK with those spec changes landing at the same time.

I had a tiny note about the $default-branch name, which is easy to see if it worked or not, after a merge. Otherwise, we can change it to the relevant string before merge. Your call, @bradshjg.

In all, this looks like future to me.

**Review by olleolleolle — APPROVED** [review]
Clear ahead!

bild

**Comment by olleolleolle** [comment]
@bradshjg Hooray! Thank you so much for your hard work in creating this!

**Pull request merged**
